src/Arara/Process/IPC/SharedMemory.php

```diff
@@ -0,0 +1,263 @@
+<?php
+
+namespace Arara\Process\IPC ;
+
+/**
+ * This class provides Interprocess communication using UNIX Shared memory
+ *
+ * Code written by Nikolay and copied into Arara/Process under the terms of
+ * the MIT license.
+ *
+ * Full credit to Nikolay for his excellent class
+ * 
+ * @license http://opensource.org/licenses/MIT
+ * @author Nikolay Bondarenko http://misterion.ru 
+ */
+class SharedMemory implements \ArrayAccess, \Countable
+{
+    /**
+     * @var resource
+     */
+    public $id;
+
+    /**
+     * @var string
+     */
+    protected $file;
+
+    /**
+     * @var int
+     */
+    protected $size;
+
+    /**
+     * @var Semaphore
+     */
+    protected $mutex;
+
+    /**
+     * Creator process pid
+     *
+     * @var int
+     */
+    protected $pid;
+
+    public function __construct($memorySize = 1024)
+    {
+        $this->pid = getmypid();
+
+        $this->size = $memorySize;
+
+        $this->file = tempnam(sys_get_temp_dir(), 's');
+        $key = ftok($this->file, 'a');
+
+        $this->id = shm_attach($key, $this->size);
+        if (false === $this->id) {
+            throw new \RuntimeException('Unable to create the shared memory segment');
+        }
+
+        //Keymapper (maps mixed to integers)
+        shm_put_var($this->id, 0, []) ;
+        $this->mutex = new Semaphore($key);
+    }
+
+    public function __destruct()
+    {
+        if ($this->pid === getmypid()) {
+            unset($this->mutex); //Destroy mutex first because of we share same System V IPC key.
+            $this->remove();
+        }
```

Comment:
This would be for similar reasons to Semaphore.  If the SharedMemory object goes out of scope or is unset manually, then that process no longer has access to the shared memory key, and therefore has no further need for the internal semaphore object.  Unsetting it in the destructor also releases any locks it may have.  

Have I missing something?  Concurrent programming is tricky to say the least. :)

D.